nixos/tests/gitlab-fetch-private.nix

```diff
@@ -0,0 +1,374 @@
+          jwsFile = pkgs.runCommand "oidcKeyBase" {} "${pkgs.openssl}/bin/openssl genrsa 2048 > $out";
+        };
+      };
+    };
+
+    client = { nodes, pkgs, ... }: {
+      imports = [
+        common/user-account.nix
+        ../modules/installer/cd-dvd/channel.nix
+      ];
+
+      virtualisation.memorySize = if pkgs.stdenv.is64bit then 512 else 256;
+      virtualisation.cores = 1;
+      virtualisation.useNixStoreImage = true;
+      virtualisation.writableStore = true;
+      virtualisation.additionalPaths = with pkgs;
+        map (pkg: writeReferencesToFile pkg.drvPath) [
+          curl gitMinimal cacert_with_extras
+        ];
+
+      security.pki.certificateFiles = [ certs.ca.cert ];
+
+      networking.extraHosts = ''
+        ${(builtins.head nodes.gitlab.networking.interfaces.eth1.ipv4.addresses).address} ${domain}
+      '';
+
+      systemd.services.nix-daemon.serviceConfig.EnvironmentFile = "/root/tokens.txt";
+    };
+  };
+
+  testScript = { nodes, ... }:
+    let
+      auth = pkgs.writeText "auth.json" (builtins.toJSON {
+        grant_type = "password";
+        username = "root";
+        password = initialRootPassword;
+      });
+
+      createUserAlice = pkgs.writeText "create-user-alice.json" (builtins.toJSON rec {
+        username = aliceUsername;
+        name = username;
+        email = "alice@localhost";
+        password = alicePassword;
+        skip_confirmation = true;
+      });
+
+      aliceAuth = pkgs.writeText "alice-auth.json" (builtins.toJSON {
+        grant_type = "password";
+        username = aliceUsername;
+        password = alicePassword;
+      });
+
+      createProjectAlice = pkgs.writeText "create-project-alice.json" (builtins.toJSON {
+        name = aliceProjectName;
+        visibility = "private";
+      });
+
+      putFile = pkgs.writeText "put-file.json" (builtins.toJSON {
+        branch = "master";
+        author_email = "author@example.com";
+        author_name = "Firstname Lastname";
+        content = "some content";
+        commit_message = "create a new file";
+      });
+
+      createTokenReadRepo = pkgs.writeText "create-project-token-read-repo.json" (builtins.toJSON {
+        id = aliceProjectId;
+        name = "test_read_repo";
+        scopes = [ "read_repository" ];
+        access_level = 20;  # "Reporter"
+      });
+
+      createTokenReadApi = pkgs.writeText "create-project-token-read-api.json" (builtins.toJSON {
+        id = aliceProjectId;
+        name = "test_read_api";
+        scopes = [ "read_api" ];
+        access_level = 20;  # "Reporter"
+      });
+
+      # The fetchers don't honor the OS trusted CAs as this would be impure.
+      # To make it accept our custom (self-signed) CA, it must be added to cacert.
+      #
+      # Fetch as if the repo was public. This must fail.
+      fetchAsPublic = pkgs.writeText "fetch-as-public.nix" ''
+        let
+          pkgs = import <nixpkgs> {};
```

Comment:
We could save a lot of rebuilds of this test by pulling the evaluation of this expression out of the VM and into the main evaluation. Instead of using it in the main build though, we can push the `.drv` into the VM. That way we don't need `<nixpkgs>` in the VM, and we won't need to rebuild this test for every little unrelated change in the repo.

So you would write the example as if you're going to use it, but instead of referencing its outputs, you reference its `drvPath` and use `builtins.unsafeDiscardOutputDependency` to remove the output from the string context (it's odd that this step is necessary; wip. It is also not unsafe.)

Instead of `nix-build ${someFile}` you could use the low level `nix-store --realise` command:

    nix-store --realise ${builtins.unsafeDiscardOutputDependency example.drvPath}

Reply:
I tried to implement your suggestion, but got "path ... is not valid" errors. Could you give me more hints what to do? My current attempt is in branch https://github.com/panicgh/nixpkgs/commits/fetchgitlab-netrc-failed_attempt

Reply:
Could be either
- nix 2.18.0, a bad release
- bug https://github.com/NixOS/nix/issues/9146

I think it's the latter, because `virtualisation.useNixStoreImage = true;` uses `exportReferencesGraph` for image building.
Could you try with `false`?
Performance difference should not be as dramatic anymore since a performance bug in qemu and/or linux (not sure) was solved. Writable store doesn't need it either.

Reply:
I tried with nix (2.13.5) and nixUnstable (2.15.2) and set `useNixStoreImage=false`. I also tried with and without `invalidateFetcherByDrvHash`, but still get `error: path '/nix/store/s1bf24apdvm6pj79bxibyymvrzhy3bmk-source' is not valid`.